dali/python/nvidia/dali/_utils/eager_util.py

```diff
@@ -0,0 +1,129 @@
+# Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import nvidia.dali.tensors as _tensors
+import nvidia.dali.types as _types
+from nvidia.dali.external_source import _prep_data_for_feed_input
+
+
+def _transform_data_to_tensorlist(data, batch_size, layout=None, device_id=None):
+    data = _prep_data_for_feed_input(data, batch_size, layout, device_id)
+
+    if isinstance(data, list):
+        if isinstance(data[0], _tensors.TensorGPU):
+            data = _tensors.TensorListGPU(data, layout or "")
+        else:
+            data = _tensors.TensorListCPU(data, layout or "")
+
+    return data
+
+
+class _Classification:
+    """Classification of data's device and if it is a batch.
+
+    Based on data type determines if data should be treated as a batch and with which device.
+    If the type can be recognized as a batch without being falsely categorized as such, it is.
+    This includes lists of supported tensor-like objects e.g. numpy arrays (the only list not
+    treated as a batch is a list of objects of primitive types), :class:`DataNodeDebug` and
+    TensorLists.
+
+    Args:
+        data: Data to be classified.
+        type_name (str): Representation of argument type (input or keyward).
+        arg_batch_size (int): Size of batch to which constant value should be scaled up. Only
+            applicable for argument inputs that are array types (e.g. numpy array). If -1 does not
+            modify the data.
+    """
+
+    def __init__(self, data, type_name, arg_batch_size=-1):
+        self.is_batch, self.device, self.data = self._classify_data(data, type_name, arg_batch_size)
+
+    @staticmethod
+    def _classify_data(data, type_name, arg_batch_size):
+        from nvidia.dali._debug_mode import DataNodeDebug
+        """Returns tuple (is_batch, device, unpacked data). """
+
+        def is_primitive_type(x):
+            return isinstance(x, (int, float, bool, str))
+
+        def classify_array_type(data):
+            if _types._is_numpy_array(data):
+                device = 'cpu'
+            elif _types._is_torch_tensor(data):
+                device = 'gpu' if data.is_cuda else 'cpu'
+                
+                # Data needs to be on CPU for batch argument inputs.
```

Comment:
Hmm, are we using this _Classification both for inputs and argument inputs? 
In some code-paths we still leave the TensorListGPU and return data tagged as 'gpu', while here we convert it to 'cpu'. 
I find this a bit convoluted.

Shouldn't we, for example, first detect what kind of data we got, and after that react accordingly:
1. I think we assumed that the inputs need to be correctly placed, so mismatch is an error - shouldn't it be the same for argument inputs
2. there is this exception of allowing GPU "constants" to be used, that in regular pipeline introduce a ConstantOp of the value - I would handle this outside of this function in explicit manner.
3. If the argument input is already a proper batch, it should already be on the CPU, shouldn't it?

Reply:
The only change to `_Classification` is when we pass something of array type (`numpy.ndarray`, `torch.Tensor` or `mxnet.ndarray.ndarray.NDArray`) then it is casted to CPU and repeated `arg_constant_len` number of times, because this is what we do in the standard pipeline when we use `ops.Constant`.

I separated those 2 cases (treated as a batch or as a constant) to different functions so it is clearer.